### PR TITLE
Add the paged library

### PR DESCRIPTION
**Pull request by npmccallum**

### CODE REVIEW

**Comment by npmccallum** [comment]
Everyone, please tell me if this is a bad idea.

Basically, I got sick of manually managing all the alignment and padding myself. I also found that alignment and padding was required in some cases, but not others, for the same data type. If this approach is sane, it allows us to decouple alignment and padding from the type definitions. All the tests pass.

File: paged/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,118 @@
+pub trait PageSize: Copy + Clone + Default {
```
Comment (steveeJ) [inline paged/src/lib.rs]: missing docstring

**Comment by npmccallum** [comment]
@steveeJ I implemented a bunch of additional, optional traits. I also added documentation. The only major change is that `Page::new()` is now `impl<S, T> From<T> for Page<S, T>`.

File: paged/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+//!
+//! struct Bar(u32, [u8; 4096]);
+//!
+//! assert_eq!(size_of::<Bar>(), size_of::<(u32, [u8; 4096])>());
+//! assert_eq!(align_of::<Bar>(), align_of::<u32>());
+//! assert_eq!(size_of::<Page<Size4k, Bar>>(), 8192);
+//! assert_eq!(align_of::<Page<Size4k, Bar>>(), 4096);
+//! ```
+//!
+//! This means that you get the alignment and padding when you need it,
```
Comment (steveeJ) [inline paged/src/lib.rs]: Would you mind pointing out (again) what _this_ refers to? I expected an aha-effect but didn't experience one, so I'm likely misunderstanding something.

File: paged/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+//! assert_eq!(size_of::<Foo>(), 8);
+//! assert_eq!(align_of::<Foo>(), align_of::<u32>());
+//!
+//! // The size and allocation of Page<Size4k, Foo> are paged.
+//! assert_eq!(size_of::<Page<Size4k, Foo>>(), 4096);
+//! assert_eq!(align_of::<Page<Size4k, Foo>>(), 4096);
+//!
+//! struct Bar(u32, [u8; 4096]);
+//!
+//! // The size and allocation of Foo are normal.
```
Comment (steveeJ) [inline paged/src/lib.rs]:
```suggestion
//! // The size and allocation of Bar are normal.
```

File: paged/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+//! assert_eq!(size_of::<Bar>(), size_of::<(u32, [u8; 4096])>());
+//! assert_eq!(align_of::<Bar>(), align_of::<u32>());
+//!
+//! // The size and allocation of Page<Size4k, Foo> are paged.
```
Comment (steveeJ) [inline paged/src/lib.rs]:
```suggestion
//! // The size and allocation of Page<Size4k, Bar> are paged.
```

File: paged/src/lib.rs (hunk 3 above)
Comment (npmccallum) [reply paged/src/lib.rs]: Fixed.

File: paged/src/lib.rs (hunk 4 above)
Comment (npmccallum) [reply paged/src/lib.rs]: Fixed.

**Pull request merged**